mypy/test/testpackages.py

```diff
@@ -0,0 +1,110 @@
+                        python_executable: str = sys.executable) -> Generator[None, None, None]:
+        """Context manager to temporarily install a package from test-data/packages/pkg/"""
+        working_dir = os.path.join(package_path, pkg)
+        install_cmd = [python_executable, '-m', 'pip', 'install', '.']
+        # if we aren't in a virtualenv, install in the
+        # user package directory so we don't need sudo
+        if not hasattr(sys, 'real_prefix') or python_executable != sys.executable:
+            install_cmd.append('--user')
+        returncode, lines = run_command(install_cmd, cwd=working_dir)
+        if returncode != 0:
+            self.fail('\n'.join(lines))
+        try:
+            yield
+        finally:
+            run_command([python_executable, '-m', 'pip', 'uninstall', '-y', pkg], cwd=package_path)
+
+    def test_get_package_dirs(self) -> None:
+        """Check that get_package_dirs works."""
+        dirs = get_package_dirs(sys.executable)
+        assert dirs
+
+    @staticmethod
+    def check_mypy_run(cmd_line: List[str],
+                       expected_out: str,
+                       expected_err: str = '',
+                       expected_returncode: int = 1) -> None:
+        """Helper to run mypy and check the output."""
+        out, err, returncode = mypy.api.run(cmd_line)
+        assert out == expected_out, err
+        assert err == expected_err, out
+        assert returncode == expected_returncode, returncode
+
+    def test_typed_package(self) -> None:
+        """Tests type checking based on installed packages.
+
+        This test CANNOT be split up, concurrency means that simultaneously
+        installing/uninstalling will break tests"""
+        with open('simple.py', 'w') as f:
+            f.write(SIMPLE_PROGRAM)
+
+        with self.install_package('typedpkg-stubs'):
+            self.check_mypy_run(
+                ['simple.py'],
+                "simple.py:4: error: Revealed type is 'builtins.list[builtins.str]'\n"
+            )
+
+        # The Python 2 tests are intentionally placed after a Python 3 test to check
+        # the package_dir_cache is behaving correctly.
+        python2 = try_find_python2_interpreter()
+        if python2:
+            with self.install_package('typedpkg-stubs', python2):
+                self.check_mypy_run(
+                    ['--python-executable={}'.format(python2), 'simple.py'],
+                    "simple.py:4: error: Revealed type is 'builtins.list[builtins.str]'\n"
+                )
+            with self.install_package('typedpkg', python2):
+                self.check_mypy_run(
+                    ['--python-executable={}'.format(python2), 'simple.py'],
+                    "simple.py:4: error: Revealed type is 'builtins.tuple[builtins.str]'\n"
+                )
+
+            with self.install_package('typedpkg', python2):
+                with self.install_package('typedpkg-stubs', python2):
```

Comment:
Could use a comma to combine these into a single `with`

Reply:
I never knew about this, interesting! However, I couldn't please flake8 about this, the next line indent wouldn't work no matter how I tried it :(